certora/specs/Gates.spec

```
@@ -0,0 +1,146 @@
+    ERC20Helper.safeTransferFrom(marketParams.loanToken, onBehalf, MorphoMarketV1Adapter.morpho, assets);
+    return (returnedShares, shares);
+}
+
+function summaryMorphoMarketV1Withdraw(MorphoMarketV1Adapter.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) returns (uint256, uint256) {
+    assert shares == 0;
+    uint256 returnedShares;
+    require (ghostStatusCanSendAssets[MorphoMarketV1Adapter.morpho], "require that MorphoMarketV1 can send assets");
+    ERC20Helper.safeTransferFrom(marketParams.loanToken, MorphoMarketV1Adapter.morpho, onBehalf, assets);
+    return (assets, returnedShares);
+}
+
+function summaryMorphoVaultV1Deposit(uint256 assets, address receiver) returns uint256 {
+    uint256 shares;
+    require (ghostStatusCanReceiveAssets[MorphoMarketV1Adapter.morpho], "require that MorphoMarketV1 can receive assets");
+    ERC20Helper.safeTransferFrom(currentContract.asset, MorphoVaultV1Adapter, MorphoMarketV1Adapter.morpho, assets);
+    return shares;
+}
+
+function summaryMorphoVaultV1Withdraw(uint256 assets, address receiver, address owner) returns uint256 {
+    uint256 shares;
+    require (ghostStatusCanSendAssets[MorphoMarketV1Adapter.morpho], "require that MorphoMarketV1 can send assets");
+    ERC20Helper.safeTransferFrom(currentContract.asset, MorphoMarketV1Adapter.morpho, MorphoVaultV1Adapter, assets);
+    return shares;
+}
+
+persistent ghost mapping(address => bool) ghostStatusCanSendShares;
+persistent ghost mapping(address => bool) ghostStatusCanReceiveShares;
+persistent ghost mapping(address => bool) ghostStatusCanSendAssets;
+persistent ghost mapping(address => bool) ghostStatusCanReceiveAssets;
+persistent ghost mapping(address => bool) ghostBalanceChangeAllowed;
+
+hook Sstore ERC20.balanceOf[KEY address user] uint256 newBalance (uint256 oldBalance) {
+    if (!canReceiveAssets(user)) {
+        ghostBalanceChangeAllowed[user] = oldBalance >= newBalance && ghostBalanceChangeAllowed[user];
+    }
+    if (!canSendAssets(user)) {
+        ghostBalanceChangeAllowed[user] = oldBalance <= newBalance && ghostBalanceChangeAllowed[user];
+    }
+}
+
+rule cantReceiveShares(env e, method f, calldataarg args, address user) {
+    require(currentContract.sharesGate != 0, "require gating to be enabled for shares");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+
+    require (!canReceiveShares(user), "require that the user under scrutiny can't receive shares");
+
+    uint256 sharesBefore = balanceOf(user);
+
+    f(e, args);
+
+    assert balanceOf(user) <= sharesBefore;
+}
+
+rule cantSendShares(env e, method f, calldataarg args, address user, uint256 shares) {
+    require(currentContract.sharesGate != 0, "require gating to be enabled for shares");
```

```suggestion
    require(currentContract.sharesGate != 0, "setup gating");
    require (!canSendShares(user), "setup gating");
```
can also be done in the rule above

---

Done in https://github.com/morpho-org/vault-v2/pull/620/commits/59a8d61de8f995cd921530c5d47ab8140e15abcd.